src/back/hlsl/writer.rs

```diff
@@ -352,29 +352,51 @@ impl<'a, W: Write> Writer<'a, W> {
         }
 
         // https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-variable-register
-        let (storage, register_ty) = match global.class {
+        let register_ty = match global.class {
             crate::StorageClass::Function => unreachable!("Function storage class"),
-            crate::StorageClass::Private => ("static ", ""),
-            crate::StorageClass::WorkGroup => ("groupshared ", ""),
-            crate::StorageClass::Uniform => ("cbuffer", "b"),
-            crate::StorageClass::Storage | crate::StorageClass::Handle => {
-                if let TypeInner::Sampler { .. } = *inner {
-                    ("", "s")
+            crate::StorageClass::Private => {
+                write!(self.out, "static ")?;
+                self.write_type(module, global.ty)?;
+                ""
+            }
+            crate::StorageClass::WorkGroup => {
+                write!(self.out, "groupshared ")?;
+                self.write_type(module, global.ty)?;
+                ""
+            }
+            crate::StorageClass::Uniform => {
+                // constant buffer declarations are expected to be inlined, e.g.
+                // `cbuffer foo: register(b0) { field1: type1; }`
+                write!(self.out, "cbuffer ")?;
+                "b"
+            }
+            crate::StorageClass::Storage => {
+                let (prefix, register) =
+                    if global.storage_access.contains(crate::StorageAccess::STORE) {
+                        ("RW", "u")
+                    } else {
+                        ("", "t")
+                    };
+                write!(self.out, "{}StructuredBuffer<", prefix)?;
```

Comment:
`write` macro inside `let _ = {}` confusing a little bit. Are you trying to avoid extra heap allocation with `format!`?

Reply:
sorry, I don't understand. This `write!` is not inside any `let _ = {}`

Reply:
Sorry, I mean a logic inside `let register_ty = match global.class {..}`.

Maybe we should split this logic ?

```rust
let register_ty = match global.class {
        crate::StorageClass::Function => unreachable!("Function storage class"),
        crate::StorageClass::Uniform => "b",
        crate::StorageClass::Storage => {
            if global.storage_access.contains(crate::StorageAccess::STORE) {
                "u"
            } else {
                "t"
            }
        }
....
```

```rust
        match global.class {
            crate::StorageClass::Function => unreachable!("Function storage class"),
            crate::StorageClass::Private => {
                write!(self.out, "static ")?;
                self.write_type(module, global.ty)?;
            }
            crate::StorageClass::Uniform => {
                // constant buffer declarations are expected to be inlined, e.g.
                // `cbuffer foo: register(b0) { field1: type1; }`
                write!(self.out, "cbuffer ")?;
            }
.....
```

Reply:
We could. I just don't want matching on the storage class multiple times

Reply:
Ok, let's keep it as is :)